File: engine/realdoor/calc.py
"""Block B — deterministic income + threshold engine (Person 2).

Implements the frozen scoring conventions:

- CH-INCOME-001: annualize *recurring gross* income by its *explicit* pay
  frequency; sum independently documented recurring sources; never infer
  undocumented income.
- HUD-MTSP-002 (via data/mtsp_2026_boston_cambridge_quincy.csv): frozen 60%
  thresholds for household sizes 1-8.  Sizes outside the table have no
  frozen threshold -> comparison "no_frozen_threshold".
- CH-DECISION-001: this module compares numbers; it never labels a person.

``annualize`` and the comparison boundary are behavior-compatible with the
starter reference (starter/src/calculate.py), extended with the
no-frozen-threshold case the starter leaves to participants.

No dollar amount, name, or household-specific branch may appear in this
file: hidden tests perturb values while retaining schemas.
"""
from __future__ import annotations

import csv
from functools import lru_cache
from pathlib import Path

from . import config
from .models import (
    Citation,
    DocumentRecord,
    IncomeSource,
    EMPLOYMENT_RATE_CONFLICT,
    GIG_INCOME_UNCORROBORATED,
    PAY_STUB_TOTAL_CONFLICT,
    UNVERIFIED_INCOME_CLAIM,
)

# Pay periods per year, keyed by the explicit frequency word used in documents
# (same table as the starter reference implementation).
FREQUENCY = {
    "weekly": 52,
    "biweekly": 26,
    "semimonthly": 24,
    "monthly": 12,
    "annual": 1,
}

BELOW_OR_EQUAL = "below_or_equal"
ABOVE = "above"
NO_FROZEN_THRESHOLD_CMP = "no_frozen_threshold"

# Two amounts are "the same money" within a cent.
CENTS = 0.01


def annualize(amount: float, frequency: str) -> float:
    """Annualize a recurring gross amount by its explicit frequency."""
    if frequency not in FREQUENCY:
        raise ValueError(f"Unsupported frequency: {frequency}")
    if amount < 0:
        raise ValueError("Amount cannot be negative")
    return round(amount * FREQUENCY[frequency], 2)


# ---------------------------------------------------------------------------
# Frozen threshold table (loaded from the pack CSV, never hardcoded)
# ---------------------------------------------------------------------------
@lru_cache(maxsize=None)
def threshold_table(csv_path: str | None = None) -> dict[int, float]:
    """household_size -> frozen 60% threshold, from the frozen MTSP CSV."""
    path = Path(csv_path) if csv_path else config.MTSP_CSV
    table: dict[int, float] = {}
    with open(path, newline="", encoding="utf-8") as handle:
        for row in csv.DictReader(handle):
            table[int(row["household_size"])] = float(row["core_challenge_threshold"])
    if not table:
        raise ValueError(f"No thresholds found in {path}")
    return table


def threshold_for(household_size: int | None) -> float | None:
    """Frozen threshold for a household size, or None outside the 1-8 table."""
    if household_size is None:
        return None
    return threshold_table().get(int(household_size))


def compare_to_threshold(annual_income: float, threshold: float | None) -> str:
    """Numeric comparison only (CH-DECISION-001: never an eligibility label)."""
    if annual_income < 0:
        raise ValueError("Income cannot be negative")
    if threshold is None:
        return NO_FROZEN_THRESHOLD_CMP
    if threshold < 0:
        raise ValueError("Threshold cannot be negative")
    return BELOW_OR_EQUAL if annual_income <= threshold else ABOVE


# ---------------------------------------------------------------------------
# Income-source derivation from document evidence
# ---------------------------------------------------------------------------
def _cite(doc: DocumentRecord, field_name: str) -> Citation | None:
    f = doc.get(field_name)
    if f is None:
        return None
    return Citation(
        document_id=doc.document_id,
        page=f.page,
        bbox=[round(float(v), 2) for v in f.bbox],
        bbox_units=f.bbox_units,
        field=f.field,
    )


def _cites(doc: DocumentRecord, *field_names: str) -> list[Citation]:
    return [c for c in (_cite(doc, n) for n in field_names) if c is not None]


def _stub_regular_amount(stub: DocumentRecord) -> tuple[float | None, bool]:
    """(recurring regular gross for one stub, conflict?).

    The recurring wage basis is regular_hours x hourly_rate: pay stubs may
    include one-off variance (overtime) in gross_pay, which is not recurring
    income under CH-INCOME-001.  A gross_pay that does not reconcile with
    the stated regular basis is a total conflict for human review.
    """
    hours = stub.value("regular_hours")
    rate = stub.value("hourly_rate")
    gross = stub.value("gross_pay")
    if hours is not None and rate is not None:
        regular = round(float(hours) * float(rate), 2)
        conflict = gross is not None and abs(float(gross) - regular) > CENTS
        return regular, conflict
    if gross is not None:
        return round(float(gross), 2), False
    return None, False


def derive_wage_source(docs: list[DocumentRecord]) -> IncomeSource | None:
    """Single recurring wage source from pay stubs + employment letter.

    Multiple stubs are repeated samples of the same recurring wage, not
    additive income.  The employment letter (weekly_hours x hourly_rate)
    corroborates the recurring basis; if no stub exists, a letter alone can
    document the wage.
    """
    stubs = [d for d in docs if d.document_type == "pay_stub"]
    letters = [d for d in docs if d.document_type == "employment_letter"]
    flags: list[str] = []
    citations: list[Citation] = []

    # (regular amount, frequency, hourly rate) per usable stub
    per_stub: list[tuple[float, str, float | None]] = []
    for stub in stubs:
        regular, conflict = _stub_regular_amount(stub)
        frequency = stub.value("pay_frequency")
        if regular is None or frequency not in FREQUENCY:
            continue
        rate = stub.value("hourly_rate")
        per_stub.append((regular, frequency,
                         None if rate is None else round(float(rate), 2)))
        citations.extend(_cites(
            stub, "regular_hours", "hourly_rate", "gross_pay", "pay_frequency"))
        if conflict and PAY_STUB_TOTAL_CONFLICT not in flags:
            flags.append(PAY_STUB_TOTAL_CONFLICT)

    # Employer letter: the hourly *rate* is the corroborated quantity.  The
    # letter's weekly hours are an approximate schedule, so hour differences
    # are not conflicts; a different rate is.
    letter_rate = letter_hours = None
    for letter in letters:
        hours = letter.value("weekly_hours")
        rate = letter.value("hourly_rate")
        if rate is not None:
            letter_rate = round(float(rate), 2)
            letter_hours = hours
            citations.extend(_cites(letter, "weekly_hours", "hourly_rate"))
            break

    if per_stub:
        stub_rates = {rate for _, _, rate in per_stub if rate is not None}
        if (letter_rate is not None and stub_rates
                and all(abs(letter_rate - r) > CENTS for r in stub_rates)):
            flags.append(EMPLOYMENT_RATE_CONFLICT)

        annuals = sorted({annualize(amount, freq) for amount, freq, _ in per_stub})
        if len(annuals) > 1:
            # Stubs disagree about the recurring wage itself.
            if PAY_STUB_TOTAL_CONFLICT not in flags:
                flags.append(PAY_STUB_TOTAL_CONFLICT)
            # Prefer a stub whose rate the employer letter corroborates,
            # else the lowest documented figure.
            corroborated = sorted(
                annualize(amount, freq) for amount, freq, rate in per_stub
                if letter_rate is not None and rate is not None
                and abs(rate - letter_rate) <= CENTS)
            annual = corroborated[0] if corroborated else annuals[0]
        else:
            annual = annuals[0]
        amount, frequency, _ = per_stub[0]
        return IncomeSource(
            kind="wages", period_amount=amount, frequency=frequency,
            annual_amount=annual, citations=citations, flags=flags)

    if letter_rate is not None and letter_hours is not None:
        weekly = round(float(letter_hours) * letter_rate, 2)
        return IncomeSource(
            kind="wages", period_amount=weekly, frequency="weekly",
            annual_amount=annualize(weekly, "weekly"),
            citations=citations, flags=flags)

    return None


def derive_benefit_sources(docs: list[DocumentRecord]) -> list[IncomeSource]:
    """One source per benefit letter: benefit amount at its explicit frequency."""
    sources = []
    for doc in (d for d in docs if d.document_type == "benefit_letter"):
        frequency = doc.value("benefit_frequency")
        amount = doc.value(f"{frequency}_benefit") if frequency else None
        if amount is None or frequency not in FREQUENCY:
            continue
        sources.append(IncomeSource(
            kind="benefit",
            period_amount=round(float(amount), 2),
            frequency=frequency,
            annual_amount=annualize(float(amount), frequency),
            citations=_cites(doc, f"{frequency}_benefit", "benefit_frequency"),
        ))
    return sources


def derive_gig_sources(docs: list[DocumentRecord]) -> list[IncomeSource]:
    """One source per gig statement: monthly gross receipts (fees are not
    deducted -- CH-INCOME-001 annualizes recurring *gross* income).  A gig
    statement alone is uncorroborated unless a gig_income_corroboration
    document is also present."""
    have_corroboration = any(
        d.document_type == "gig_income_corroboration" for d in docs)
    sources = []
    for doc in (d for d in docs if d.document_type == "gig_statement"):
        receipts = doc.value("gross_receipts")
        if receipts is None or doc.value("statement_month") is None:
            continue
        flags = [] if have_corroboration else [GIG_INCOME_UNCORROBORATED]
        sources.append(IncomeSource(
            kind="gig",
            period_amount=round(float(receipts), 2),
            frequency="monthly",                    # one statement month
            annual_amount=annualize(float(receipts), "monthly"),
            citations=_cites(doc, "gross_receipts", "statement_month"),
            flags=flags,
        ))
    return sources


def derive_self_declared_sources(docs: list[DocumentRecord]) -> list[IncomeSource]:
    """Self-declared amounts (e.g. on an application summary) are never
    counted as documented income; they surface as UNVERIFIED_INCOME_CLAIM."""
    sources = []
    for doc in (d for d in docs if d.document_type == "application_summary"):
        amount = doc.value("declared_income")
        frequency = doc.value("declared_income_frequency") or "annual"
        if amount is None:
            continue
        sources.append(IncomeSource(
            kind="self_declared",
            period_amount=round(float(amount), 2),
            frequency=frequency,
            annual_amount=0.0,
            citations=_cites(doc, "declared_income"),
            flags=[UNVERIFIED_INCOME_CLAIM],
            counted=False,
        ))
    return sources


def derive_income_sources(docs: list[DocumentRecord]) -> list[IncomeSource]:
    """All documented recurring sources for one household's documents."""
    sources: list[IncomeSource] = []
    wage = derive_wage_source(docs)
    if wage is not None:
        sources.append(wage)
    sources.extend(derive_benefit_sources(docs))
    sources.extend(derive_gig_sources(docs))
    sources.extend(derive_self_declared_sources(docs))
    return sources


def total_annualized_income(sources: list[IncomeSource]) -> float:
    """Sum of counted, independently documented sources (CH-INCOME-001)."""
    return round(sum(s.annual_amount for s in sources if s.counted), 2)

File: frontend/lib/pipeline/calc.ts
// Deterministic challenge math — ported from the starter pack's calculate.py.
// NOT an eligibility engine: it annualizes income, compares to a published
// threshold, and derives file-level readiness. Money is handled in INTEGER CENTS
// (never floats) to keep the math exact.

import type {
  DocumentRecord,
  DocumentType,
  ExtractedField,
  Comparison,
  DisplayStatus,
  ReadinessResult,
  ReviewReason,
  ReviewReasonCode,
} from "@/lib/pipeline/types";
import type { PayFrequency } from "@/lib/pipeline/types";
import { thresholdForSize } from "@/lib/data/mtsp2026";

export const FREQUENCY: Record<PayFrequency, number> = {
  weekly: 52,
  biweekly: 26,
  semimonthly: 24,
  monthly: 12,
  annual: 1,
};

/** Event/"today" date for the challenge; the 60-day currency window ends here. */
export const EVENT_DATE = "2026-07-18";
export const CURRENCY_WINDOW_START = "2026-05-19"; // >= this is "current"
const LOW_CONFIDENCE = 0.75;

/** Parse a dollar string/number into integer cents (rounded). */
export function toCents(dollars: string | number): number {
  const n = typeof dollars === "number" ? dollars : Number(String(dollars).replace(/[^0-9.-]/g, ""));
  if (!Number.isFinite(n)) return 0;
  return Math.round(n * 100);
}

/** Annualize an amount (in cents) by its pay frequency; rejects unknown freq / negatives. */
export function annualizeCents(amountCents: number, frequency: PayFrequency): number {
  const mult = FREQUENCY[frequency];
  if (mult === undefined) throw new Error(`Unsupported frequency: ${frequency}`);
  if (amountCents < 0) throw new Error("Amount must be non-negative");
  return amountCents * mult;
}

/** Inclusive boundary: <= threshold => below_or_equal (pinned by the organizer test). */
export function compareToThreshold(annualCents: number, thresholdCents: number | null): Comparison {
  if (thresholdCents === null) return "no_frozen_threshold";
  return annualCents <= thresholdCents ? "below_or_equal" : "above";
}

export function thresholdCentsForSize(size: number | null): number | null {
  if (size === null) return null;
  const dollars = thresholdForSize(size);
  return dollars === null ? null : dollars * 100;
}

export function formatMoneyCents(cents: number, locale: string = "en-US"): string {
  return (cents / 100).toLocaleString(locale, {
    style: "currency",
    currency: "USD",
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  });
}

/** Plain number with 2 decimals (for submission.json annualized_income). */
export function centsToDollars(cents: number): number {
  return Math.round(cents) / 100;
}

/** One documented income source in the formula breakdown. `counted: false`
 *  means it corroborates an already-counted source and is NOT added (§8). */
export type IncomeSourceEntry = {
  fieldId: string;
  documentId: string;
  key: string;
  documentType: DocumentType;
  periodCents: number;
  frequency: PayFrequency;
  annualCents: number;
  counted: boolean;
};

/**
 * Derive the income sources from CONFIRMED/CORRECTED fields only.
 * Domain law (§8): multiple pay stubs from the same recurring wage source
 * corroborate — the stub with the latest pay_date is counted once, the others
 * confirm it. Benefit letters and gig statements each stand on their own
 * (gig receipts are monthly, ×12).
 */
export function deriveIncomeSources(
  documents: readonly DocumentRecord[],
  fields: readonly ExtractedField[],
): IncomeSourceEntry[] {
  const resolvedIncome = fields.filter((f) => f.isIncome && isResolved(f));
  const docType = (id: string): DocumentType =>
    documents.find((d) => d.id === id)?.documentType ?? "unknown";

  // Wages: pick the pay stub with the latest confirmed pay_date as THE source.
  const stubs = resolvedIncome.filter((f) => docType(f.documentId) === "pay_stub");
  const payDate = (f: ExtractedField): number => {
    const raw = fields.find((x) => x.documentId === f.documentId && x.key === "pay_date")?.value;
    const parsed = Date.parse(raw ?? "");
    return Number.isNaN(parsed) ? 0 : parsed;
  };
  const countedStubId =
    stubs.length > 0 ? [...stubs].sort((a, b) => payDate(b) - payDate(a))[0].id : undefined;

  // §8: a stub's recurring basis is regular_hours × hourly_rate when both are
  // confirmed (the internally consistent figure — on a printed-total conflict
  // this basis wins, never the printed gross, never an average or max).
  // Stubs without hours/rate fall back to the printed gross.
  const stubBasisCents = (f: ExtractedField): number => {
    const rate = fields.find((x) => x.documentId === f.documentId && x.key === "hourly_rate");
    const hours = fields.find((x) => x.documentId === f.documentId && x.key === "regular_hours");
    if (rate && hours && isResolved(rate) && isResolved(hours)) {
      const hoursNum = Number(hours.value);
      const rateCents = toCents(rate.value);
      if (Number.isFinite(hoursNum) && hoursNum > 0 && rateCents > 0) {
        return Math.round(rateCents * hoursNum);
      }
    }
    return toCents(f.value);
  };

  const entries: IncomeSourceEntry[] = [];
  for (const f of resolvedIncome) {
    const documentType = docType(f.documentId);
    const frequency = f.incomeFrequency ?? "monthly";
    const periodCents = documentType === "pay_stub" ? stubBasisCents(f) : toCents(f.value);
    // Defense in depth: a negative/unparseable amount is invalid input, never a
    // source (the UI also rejects negative corrections at the boundary) — and
    // it must never reach annualizeCents, which throws on negatives.
    if (periodCents < 0) continue;
    const counted = documentType !== "pay_stub" || f.id === countedStubId;
    entries.push({
      fieldId: f.id,
      documentId: f.documentId,
      key: f.key,
      documentType,
      periodCents,
      frequency,
      annualCents: annualizeCents(periodCents, frequency),
      counted,
    });
  }

  // No pay stub at all: an employment letter's confirmed weekly_hours ×
  // hourly_rate stands as the recurring wage source (weekly). With stubs
  // present the letter only corroborates — it is never added on top.
  if (stubs.length === 0) {
    const letterDocs = documents.filter((d) => d.documentType === "employment_letter");
    const letterDate = (docId: string): number => {
      const raw = fields.find((x) => x.documentId === docId && x.key === "document_date")?.value;
      const parsed = Date.parse(raw ?? "");
      return Number.isNaN(parsed) ? 0 : parsed;
    };
    const candidates = letterDocs
      .map((d) => {
        const hours = fields.find(
          (x) => x.documentId === d.id && x.key === "weekly_hours" && isResolved(x),
        );
        const rate = fields.find(
          (x) => x.documentId === d.id && x.key === "hourly_rate" && isResolved(x),
        );
        return hours && rate ? { doc: d, hours, rate } : null;
      })
      .filter((x): x is NonNullable<typeof x> => x !== null)
      .sort((a, b) => letterDate(b.doc.id) - letterDate(a.doc.id));
    const pick = candidates[0];
    if (pick) {
      const hoursNum = Number(pick.hours.value);
      const rateCents = toCents(pick.rate.value);
      if (Number.isFinite(hoursNum) && hoursNum > 0 && rateCents > 0) {
        const periodCents = Math.round(rateCents * hoursNum);
        entries.push({
          fieldId: pick.hours.id,
          documentId: pick.doc.id,
          key: "letter_wage",
          documentType: "employment_letter",
          periodCents,
          frequency: "weekly",
          annualCents: annualizeCents(periodCents, "weekly"),
          counted: true,
        });
      }
    }
  }

  return entries;
}

/** Gross annual income (cents): the sum of COUNTED sources only. */
export function computeGrossAnnualCents(
  documents: readonly DocumentRecord[],
  fields: readonly ExtractedField[],
): number {
  return deriveIncomeSources(documents, fields)
    .filter((s) => s.counted)
    .reduce((sum, s) => sum + s.annualCents, 0);
}

function isExpired(dateStr: string): boolean {
  // "current" if dated on/after the 60-day window start (challenge convention).
  // A "YYYY-MM" statement month counts as its LAST covered day (engine parity).
  const text = dateStr.trim();
  let d: number;
  if (/^\d{4}-\d{2}$/.test(text)) {
    const year = Number(text.slice(0, 4));
    const month = Number(text.slice(5, 7));
    d = Date.UTC(year, month, 0); // day 0 of next month = last day of `month`
  } else {
    d = Date.parse(text);
  }
  if (Number.isNaN(d)) return false;
  return d < Date.parse(CURRENCY_WINDOW_START);
}

function isResolved(f: ExtractedField): boolean {
  return f.reviewStatus === "confirmed" || f.reviewStatus === "corrected" || f.reviewStatus === "renter_entered";
}

/**
 * File-level readiness. Blocking codes drive the status; missing documents are
 * informational (they raise Errors but do not block) — per CLAUDE.md §6.
 */
export function computeReadiness(
  documents: readonly DocumentRecord[],
  fields: readonly ExtractedField[],
  missingRequired: readonly string[],
  householdSize: number | null,
): ReadinessResult {
  const reasons: ReviewReason[] = [];

  // Evidence checks run on CONFIRMED values only — an unconfirmed extraction is
  // "pending", not yet an inconsistency. Once the renter confirms (or corrects)
  // the involved values, the checks below fire independently of the rest.

  // Frozen 60-day currency convention over EVERY dated evidence document
  // (mirrors the engine's DATE_FIELD map; one reason per expired code).
  const DATE_FIELD: Partial<Record<DocumentType, string>> = {
    employment_letter: "document_date",
    benefit_letter: "document_date",
    pay_stub: "pay_date",
    gig_statement: "statement_month",
  };
  const EXPIRED_CODE: Partial<Record<DocumentType, ReviewReasonCode>> = {
    employment_letter: "EMPLOYMENT_LETTER_EXPIRED",
    benefit_letter: "BENEFIT_LETTER_EXPIRED",
    pay_stub: "PAY_STUB_EXPIRED",
    gig_statement: "GIG_STATEMENT_EXPIRED",
  };
  const expiredSeen = new Set<ReviewReasonCode>();
  for (const doc of documents) {
    const dateKey = DATE_FIELD[doc.documentType];
    const code = EXPIRED_CODE[doc.documentType];
    if (!dateKey || !code || expiredSeen.has(code)) continue;
    const dateField = fields.find((f) => f.documentId === doc.id && f.key === dateKey);
    if (dateField && isResolved(dateField) && isExpired(dateField.value)) {
      expiredSeen.add(code);
      reasons.push({ code, blocking: true, documentId: doc.id });
    }
  }

  // Gig income corroboration — flagged once its receipts are confirmed.
  for (const doc of documents) {
    if (doc.documentType !== "gig_statement") continue;
    const receipts = fields.find((f) => f.documentId === doc.id && f.key === "gross_receipts");
    if (receipts && isResolved(receipts)) {
      reasons.push({ code: "GIG_INCOME_UNCORROBORATED", blocking: true, documentId: doc.id });
    }
  }

  // Employer letter corroborates the RATE: a confirmed letter rate differing
  // from every confirmed stub rate (beyond a cent) is a conflict; differing
  // hours are just an approximate schedule, never a conflict.
  const stubRates = documents
    .filter((d) => d.documentType === "pay_stub")
    .map((d) => fields.find((f) => f.documentId === d.id && f.key === "hourly_rate"))
    .filter((f): f is ExtractedField => Boolean(f && isResolved(f)))
    .map((f) => toCents(f.value))
    .filter((cents) => cents > 0);
  const letterRateField = documents
    .filter((d) => d.documentType === "employment_letter")
    .map((d) => fields.find((f) => f.documentId === d.id && f.key === "hourly_rate"))
    .find((f) => f && isResolved(f));
  if (letterRateField && stubRates.length > 0) {
    const letterRate = toCents(letterRateField.value);
    if (letterRate > 0 && stubRates.every((r) => Math.abs(r - letterRate) > 1)) {
      reasons.push({
        code: "EMPLOYMENT_RATE_CONFLICT",
        blocking: true,
        documentId: letterRateField.documentId,
      });
    }
  }

  // Self-declared amounts are never counted as documented income.
  for (const doc of documents) {
    if (doc.documentType !== "application_summary") continue;
    const declared = fields.find((f) => f.documentId === doc.id && f.key === "declared_income");
    if (declared && isResolved(declared)) {
      reasons.push({ code: "UNVERIFIED_INCOME_CLAIM", blocking: true, documentId: doc.id });
    }
  }

  // Threshold applicability + income evidence — only once nothing is pending,
  // so "not confirmed yet" is never mislabelled as "missing".
  const nothingPending = documents.length > 0 && fields.every((f) => isResolved(f));
  if (nothingPending) {
    if (householdSize === null) {
      reasons.push({ code: "MISSING_HOUSEHOLD_SIZE", blocking: true });
    } else if (thresholdForSize(householdSize) === null) {
      reasons.push({ code: "NO_FROZEN_THRESHOLD", blocking: true });
    }
    const counted = deriveIncomeSources(documents, fields).some((s) => s.counted);
    if (!counted) {
      reasons.push({ code: "MISSING_INCOME_EVIDENCE", blocking: true });
    }
  }

  // Pay-stub internal conflict: stated gross vs hourly_rate * regular_hours,
  // once all three values are confirmed.
  for (const doc of documents) {
    if (doc.documentType !== "pay_stub") continue;
    const gross = fields.find((f) => f.documentId === doc.id && f.key === "gross_pay");
    const rate = fields.find((f) => f.documentId === doc.id && f.key === "hourly_rate");
    const hours = fields.find((f) => f.documentId === doc.id && f.key === "regular_hours");
    if (gross && rate && hours && isResolved(gross) && isResolved(rate) && isResolved(hours)) {
      const stated = toCents(gross.value);
      const computed = Math.round(toCents(rate.value) * Number(hours.value)) / 1; // cents
      if (stated !== computed) {
        reasons.push({ code: "PAY_STUB_TOTAL_CONFLICT", blocking: true, documentId: doc.id });
      }
    }
  }

  // Unresolved / low-confidence extracted fields.
  const unresolved = fields.filter((f) => !isResolved(f));
  if (unresolved.length > 0) {
    reasons.push({ code: "UNCONFIRMED_FIELDS", blocking: true, detail: String(unresolved.length) });
  }
  const lowConf = unresolved.filter((f) => f.confidence < LOW_CONFIDENCE);
  if (lowConf.length > 0) {
    reasons.push({ code: "LOW_CONFIDENCE_FIELDS", blocking: false, detail: String(lowConf.length) });
  }

  // Missing required documents — informational only (do NOT block).
  for (const docType of missingRequired) {
    reasons.push({ code: "MISSING_REQUIRED_DOCUMENT", blocking: false, detail: docType });
  }

  const status = reasons.some((r) => r.blocking) ? "NEEDS_REVIEW" : "READY_TO_REVIEW";
  return { status, reasons };
}

/** Extracted fields the renter has not yet confirmed/corrected/entered. */
export function countUnresolved(fields: readonly ExtractedField[]): number {
  return fields.filter((f) => !isResolved(f)).length;
}

/**
 * Progress-aware display status, first match wins. UI vocabulary only: the
 * exported readiness stays the organizer's binary and is computed separately
 * in computeReadiness().
 */
export function deriveDisplayStatus(args: {
  documentCount: number;
  busy: boolean;
  locked: boolean;
  unresolvedCount: number;
  reasons: readonly ReviewReason[];
  missingRequiredCount: number;
}): DisplayStatus {
  if (args.documentCount === 0) return "NOT_STARTED";
  if (args.busy) return "PROCESSING";
  if (args.locked) return "PACKET_LOCKED";
  // Detected inconsistencies outrank pending confirmations (red beats yellow).
  if (args.reasons.some((r) => r.blocking && r.code !== "UNCONFIRMED_FIELDS")) {
    return "EVIDENCE_ISSUES";
  }
  if (args.unresolvedCount > 0) return "AWAITING_CONFIRMATION";
  if (args.missingRequiredCount > 0) return "DOCUMENTS_MISSING";
  return "READY";
}

/**
 * Errors = detected inconsistencies / rule flags on confirmed values.
 * Independent of confirmation progress: unconfirmed fields and missing
 * documents are "pending"/informational, never errors.
 */
export function computeErrorCount(reasons: readonly ReviewReason[]): number {
  return reasons.filter((r) => r.blocking && r.code !== "UNCONFIRMED_FIELDS").length;
}

export { LOW_CONFIDENCE };
